### PR TITLE
eviction by process number

### LINKED ISSUE #89284: Kubelet Out-of-Pid eviction should rank based on pids used
**What would you like to be added**:
Enhance the kubelet's eviction algorithm when handling out-of-pid scenarios to evict the pod using the most pids.

**Why is this needed**:
Currently, the kubelet ranks pods for out-of-pid eviction only by PodPriority.  When there are many pods of the same priority, it ranks them randomly, which is not ideal.  In the case of a fork bomb, the kubelet prevents the node from running out of pids, but may evict many innocent pods before choosing the fork bomb.

**How to implement this**

cAdvisor already collects process stats:  https://github.com/google/cadvisor/blob/master/info/v1/container.go#L814

The kubelet already enables process stats: 
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/server/server.go#L343

These metrics just need to be used to make eviction decisions here:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/eviction/helpers.go#L642
Notice that pid metrics are only ranked by priority, whereas disk and memory are ranked by (1) whether usage > requests, (2) priority, (3) usage - requests.

The eviction manager's decisions are made based on the Summary API:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/stats/v1alpha1/types.go

We need to add ProcessMetrics to the PodStats struct.  We can start with number of processes.  Something like:

```golang
type ProcessStats struct {
	// Number of processes
	// +optional
	ProcessCount *uint64 `json:"process_count,omitempty"`
}

// PodStats holds pod-level unprocessed sample stats.
type PodStats struct {
        ...
	// Stats pertaining to processes.
	// +optional
	ProcessStats *ProcessStats `json:"process_stats,omitempty"`
}
```

The summary API is constructed from the cadvisorInfo, and the https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/stats/helper.go#L143 has methods for extracting particular metrics from the cAdvisorInfo structs.

This needs to be set when constructing the summary API in both the [cri_stats_provider](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/stats/cri_stats_provider.go) and the cadvisor_stats_provider)[https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/stats/cadvisor_stats_provider.go]

The eviction manager just needs to be able to take a [PodStats](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/stats/v1alpha1/types.go#L95), and get the usage, like [memoryUsage(...](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/eviction/helpers.go#L323), then to implement a compare function for pids, like [memory(...](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/eviction/helpers.go#L546)

There is no need to implement an "exceedsRequests" function for pids, as there are no per-pod pid requests.  There is also no need to add container-level metrics to the summary API.  Pod-level metrics should be sufficient for our purposes.

**Testing**

To test the ranking, update the [PriorityPidEvictionOrdering](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/eviction_test.go#L398) node e2e test.  Right now, it only has two pods, one that is malicious and low-priority, and one that is innocent and high-priority.  It relies on the priority ranking to ensure the malicious pod is evicted first.  We want to add a third pod, which is a high-priority malicious pod, with the same priority as the innocent pod.  This will test that the pod with higher PID usage is evicted first.

/sig node
/kind feature
/priority important-longterm
/help

### PR DESCRIPTION
**Pull request by gongguan**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes [#89284](https://github.com/kubernetes/kubernetes/issues/89284)

**Special notes for your reviewer**:
Enhance the kubelet's eviction algorithm when handling out-of-pid scenarios to evict the pod using the most pids.

**Does this PR introduce a user-facing change?**:

```release-note
None
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: test/e2e_node/eviction_test.go [hunk 1]
```diff
@@ -422,15 +422,20 @@ var _ = framework.KubeDescribe("PriorityPidEvictionOrdering [Slow] [Serial] [Dis
...
+				evictionPriority: 2,
+				pod:              pidConsumingPod("fork-bomb-container-with-low-priority", 12000),
 			},
 			{
 				evictionPriority: 0,
 				pod:              innocentPod(),
 			},
+			{
+				evictionPriority: 1,
+				pod:              pidConsumingPod("fork-bomb-container-with-high-priority", 1200),
```
Comment (dashpole) [inline test/e2e_node/eviction_test.go]: I believe this needs to be 12000

**Review by dashpole — COMMENTED** [review]
have you been able to run the pid eviction e2e test?  It is not run as part of presubmits.

Since this is an API change, i'm going to raise it at sig-node just to make sure everyone is on-board.

**Comment by mattjmcnaughton** [comment]
@dashpole @gongguan given this involves an API change, do we think a KEP could be helpful here? Or not necessary?

(Disclosure, I missed the sig-node meeting today, so may have missed what was discussed :) )

**Comment by dashpole** [comment]
Sig-node should be tomorrow morning.  This falls under https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190129-pid-limiting.md, but may require us to fill in a few more details

**Comment by gongguan** [comment]
> have you been able to run the pid eviction e2e test? It is not run as part of presubmits.

Sorry, do you mean run pid eviction e2e test locally because it hasn't been tested by automatic check?

File: test/e2e_node/eviction_test.go (hunk 1 above)
Comment (gongguan) [reply test/e2e_node/eviction_test.go]: addressed

**Comment by gongguan** [comment]
run `./kubetest --provider=local --test --test_args=--ginkgo.focus="NodeFeature:Eviction"` locally, but all tests skipped. Um, is there something wrong in my command?

**Comment by dashpole** [comment]
I think you have to explicitly set ginkgo.skip="" since serial tests are skipped by default.  I would also suggest Specifically focusing the PidEviction tests, as there are many tests that will be run otherwise

**Comment by gongguan** [comment]
Failed to test node-e2e cases locally, see [#89431](https://github.com/kubernetes/kubernetes/issues/89431)

**Comment by dashpole** [comment]
You will want `make test-e2e-node FOCUS=PidEviction SKIP=''`  You can add `REMOTE=true` if you want it to tun on a cloud provider instead of locally

**Comment by derekwaynecarr** [comment]
discussed https://github.com/kubernetes/enhancements/pull/1631 in sig node and agreed to proceed.

thanks for improving this feature!

**Comment by dashpole** [comment]
I checked out your PR and ran it on GCP, and it passed.  Good work!

**Comment by dashpole** [comment]
looks great!
/lgtm
/approve

**Comment by dashpole** [comment]
/assign @Random-Liu 
can you approve for the e2e node changes?

**Comment by xujyan** [comment]
@dashpole @Random-Liu slightly tangent: we can implement eviction by file descriptor count in the same way right?

**Comment by dashpole** [comment]
@xujyan That makes sense to me.  There is one more improvement I would like to make to pid eviction before we move on to other resources.  That is to implement eviction for node allocatable for pids.

**Pull request merged**